### PR TITLE
refactor: KOS and preprocessing traits

### PR DESCRIPTION
**Pull request by sinui0**
This PR adds traits for the preprocessing model and also updates KOS to implement random OT for any `Standard: Distribution<T>` type.

`Allocate` trait rationale
---

It is difficult to properly model preprocessing in the case where a functionality is shared by multiple others. One approach would be to update the `preprocess` function so that it has a `count` argument then use some sort of async `Barrier` for synchronization. That approach doesn't work in the case where multiplexing isn't available (only 1 context).

The approach I settled on was to decouple reserving capacity step from execution of the preprocessing. The implemented model is 2 steps:

1. A functionality is injected as a sub-protocol into other protocols. Those protocols themselves implement `Allocate`, which allocate what they need from the sub-protocol. This process does not requires synchronization.
2. Once all protocols have allocated what they need, `Preprocess` is called which actually executes the preprocessing all batched together.

Modeling it this way works for both cases where a functionality is "shared" or owned directly by another protocol.

Side note: Once we have SoftSpoken which can be extended multiple times, we should be able to just share the base OT and perform OT extensions in parallel instead of this batching. This will remove the need for shared state synchronization and be more performant in terms of latency and compute.

### CODE REVIEW

File: crates/mpz-common/src/lib.rs [hunk 1]
```diff
@@ -23,12 +23,29 @@ pub mod ideal;
...
+pub trait Allocate {
+    /// Allocates `count` capacity.
+    fn alloc(&mut self, count: usize);
+}
+
+/// A functionality in the pre-processing model.
+#[async_trait]
+pub trait Preprocess<Ctx>: Allocate {
+    /// Error type.
+    type Error;
```
Comment (th4s) [inline crates/mpz-common/src/lib.rs]: Can we add a `std::error::Error` trait bound here? That usually makes it easier to work with.

File: crates/mpz-share-conversion/src/sender.rs [hunk 2]
```diff
@@ -1,29 +1,66 @@
...
 use rand::thread_rng;
 use serio::{Deserialize, Serialize, SinkExt};
 use std::marker::PhantomData;
 
 /// Sender for share conversion.
+#[derive(Debug)]
 pub struct ShareConversionSender<T, F> {
     ole_sender: T,
-    phantom: PhantomData<F>,
+    _pd: PhantomData<F>,
```
Comment (th4s) [inline crates/mpz-share-conversion/src/sender.rs:15]: Let's have a consistent name for `PhantomData` for sender and receiver.

**Review by th4s — APPROVED** [review]
Very nice :rocket:

**Review by themighty1 — APPROVED** [review]
ACK

**Pull request merged**